### PR TITLE
reduce cyclocomp of lint_parse_error() to 5

### PR DESCRIPTION
**Pull request by AshesITR**
Part of #1186 

split into helpers for different cases

 - lint_parse_error_r43 (CC 6)
 - lint_parse_error_r42 (CC 2)
 - lint_parse_error_nonstandard (CC 14)

### CODE REVIEW

File: R/get_source_expressions.R [hunk 1]
```diff
@@ -124,229 +124,237 @@ get_source_expressions <- function(filename, lines = NULL) {
...
-    column_number <- nchar(line)
-  } else {
-    line <- source_expression$lines[[line_number]]
+  message_info <- re_matches(e$message,
+                             rex(single_quotes, capture(name = "name", anything), single_quotes,
+                                 anything,
+                                 double_quotes, capture(name = "starting", anything), double_quotes))
+
+  loc <- re_matches(source_expression$content, rex(message_info$starting), locations = TRUE)
+  line_location <- loc[!is.na(loc$start) & !is.na(loc$end),]
```
Comment (MichaelChirico) [inline R/get_source_expressions.R:318]: de-lint

File: R/get_source_expressions.R [hunk 2]
```diff
@@ -124,229 +124,237 @@ get_source_expressions <- function(filename, lines = NULL) {
...
+  if (nrow(line_location) == 0L) {
+    if (grepl("attempt to use zero-length variable name", e$message, fixed = TRUE)) {
+      # empty symbol: ``, ``(), ''(), ""(), fun(''=42), fun(""=42), fun(a=1,""=42)
+      loc <- re_matches(source_expression$content,
+                        rex("``" %or%
+                              list(or("''", '""'), any_spaces, "(") %or%
+                              list(or("(", ","), any_spaces, or("''", '""'), any_spaces, "=")),
+                        options = "multi-line",
+                        locations = TRUE)
+      loc <- loc[!is.na(loc$start) & !is.na(loc$end),]
```
Comment (MichaelChirico) [inline R/get_source_expressions.R:329]: de-lint

File: R/get_source_expressions.R [hunk 3]
```diff
@@ -124,229 +124,237 @@ get_source_expressions <- function(filename, lines = NULL) {
...
+      if (nrow(loc) > 0L) {
+        line_location <- loc[1L,]
```
Comment (MichaelChirico) [inline R/get_source_expressions.R:331]: de-lint

File: R/get_source_expressions.R [hunk 4]
```diff
@@ -124,229 +124,237 @@ get_source_expressions <- function(filename, lines = NULL) {
...
       )
-      l <- as.integer(rx_match$line)
-      # Sometimes the parser "line" runs one past the last line
-      l <- pmin(l, length(source_expression$lines))
+    )
+  }
 
-      msg <- paste0(rx_match$msg_1, if (nzchar(rx_match$msg_2)) " ", rx_match$msg_2, ".")
-      substr(msg, 1L, 1L) <- toupper(substr(msg, 1L, 1L))
+  # Hand-crafted regex to parse all error messages generated by the R parser code for R < 4.3.
```
Comment (MichaelChirico) [inline R/get_source_expressions.R:262]: make this into roxygen formatting with noRd

File: R/get_source_expressions.R [hunk 5]
```diff
@@ -124,229 +124,237 @@ get_source_expressions <- function(filename, lines = NULL) {
...
-          line_number = l,
-          column_number = 1L,
-          type = "error",
-          message = "Invalid multibyte character in parser. Is the encoding correct?",
-          line = source_expression$lines[[l]]
-        )
+  lint_parse_error_nonstandard(e, source_expression)
+}
+
+lint_parse_error_r43 <- function(e, source_expression) {
```
Comment (MichaelChirico) [inline R/get_source_expressions.R:153]: the naming is a bit obscure, we should add internal roxygen comments to these helpers with a little explanation

Comment (AshesITR) [reply R/get_source_expressions.R:153]: will do

**Review by MichaelChirico — APPROVED** [review]

**Pull request merged**